### ISSUE TITLE
FR: thumbnail strip with labels for pb-facsimile

### ISSUE THREAD

**Issue by tuurma** [eeditiones/tei-publisher-components#85]
extend pb-facsimile reference strip feature with a way to specify a label for each of the thumbnails, like it is done in e.g. [Ginzberg edition](https://www.ginzberg.ethz.ch/05_Vol_03_folder_01_001r.xml)

<img width="881" alt="image" src="https://user-images.githubusercontent.com/449468/162000158-03e33921-97cd-4ae9-b708-cc0a40248162.png">

**Comment by line-o** [comment]
Notes from our discusssion

- Vanilla JS example: https://www.ginzberg.ethz.ch/05_Vol_03_folder_01_001r.xml
- OSD plugin example: https://rrss.manuscripta.zrc-sazu.si/rrss_ms_002

Strip shows all facsimiles given to it

List of thumbnails

Communicates with OSD

Must allow to pass in additional metadata

Must be placeable at a different location on the page

- Outside the actual facsimile view
- Below, over, left-of, right-of the view

`<pb-facsimile base-uri=“{base}”>`

- Sends an event to 

`<pb-facsimile-strip [base-uri=“{base}”] emit=“{channel}” subscribe=“{channel}”>`

- emits and event on emit-channel

`<pb-facs-link facs=“relative url base of the ”>inline content<span slot=“label”></span></pb-facs-link>`

We were looking at two use cases:

Use case A: 1:1 relationship - text changes each time a different thumbnail

Use Case B: 1:n relationship - text changes only when clicked thumbnail is not part of current context

Implementation

Copy <pb-facsimile> demo

**Comment by tuurma** [comment]
Use cases:

## Source of facsimiles and their labels

### facs-strip as an overview: all facsimiles available in the current transcription fragment

* pb-view or pb-load with pb-facs-link elements serve as a source for facsimiles list for pb-facsimile and pb-facs-strip
* selecting a thumbnail from the strip emits an event to display the chosen image in the pb-facsimile; other views are not affected

* Each document in an edition may consist of one or more pages. Transcription presents the full text but user can switch between image pages via the thumbnail strip.*

### facs-strip as edition navigation

* list of facsimiles comes from an api endpoint or an arbitrary pb-view/pb-load and may include links to facsimiles not referenced in the currently presented portion of the text
* selecting an image behaves like pb-link: refreshes the views without reloading the entire page

*Let's imagine an edition organized in lengthy chapters (like Bible). The document view is organized with facing transcription/facsimile panels and page by page navigation. Nevertheless, to assist navigation, thumbnails strip with opening pages for each chapter is available at all times at the bottom.*

## Organization of the strip

* layout of the image tiles: horizontal vs vertical
* positioning of the label relative to the image: top, bottom (perhaps also right and left?)
* text-direction for the label: ltr, rtl (unless this can be styled from the outside)

css properties:
* label styles: font, color, size, horizontal and vertical alignment
* img: width and height

## other concerns

* avoid presenting duplicate thumbnails, even if there are multiple pb-facs-links for the same page (perhaps via a selector property on pb-facs-strip)
* some sort of limitation to prevent accidentally handling huge image sets?
* ideally removing the limitation that images can only come from a pb-view, perhaps do it via a selector: 
   - source="api/thumbnails"
   - selector=".view .thumbs" (perhaps tricky as it would have to retrieve links from pb-view/pb-load's shadow dom

**Comment by JoernT** [comment]
further i suggest to create some usage examples - how will the element used on the page?

e.g.

```
<pb-facs>
    <pb-facs-strip slot="top"></pb-facs-strip>
</pb-facs>
```

or
```
<pb-facs id="my-facs">

</pb-facs>
 <pb-facs-strip viewer="my-facs"></pb-facs-strip>
```

i guess containment (first example) is more sensible and eases the positioning. However the positioning would be rather an enhancement to `<pb-facs>` providing appropriate slots (left, right, top, bottom) that a part of `<pb-facs-strip`.

Hope that makes sense

**Comment by line-o** [comment]
> name the component 'pb-facs-thumbs'

Naming thing is hard, so we should think about a good name. I think your suggestion does make sense but I also think that `strip` as in comic-strip is well understood as a series of images.

I would think that `thumbs` is better if we want to add a grid view.

**Comment by line-o** [comment]
```
<pb-facs>
    <pb-facs-strip slot="top"></pb-facs-strip>
</pb-facs>
```
Very interesting suggestion @JoernT - I like it for easy placement.
I believe the goal is to also be able to put the thumbnails view anywhere on the page. 
Definitely the `pb-facs-strip` can exist on a page of its own without a `pb-facs-viewer` as a means of navigating a text.
@tuurma @wolfgangmm did I get that right?

**Comment by JoernT** [comment]
Even if contained in a slot it could be positioned anywhere with position=fixed.

And i agree that the components should at least be thought of as being independent of pb-facs

**Comment by tuurma** [comment]
Typical markup when the list of thumbnails is retrieved from the `pb-view`, so the same as the list that `pb-facsimile` gets
         
         ```
         <main>
            <pb-view id="view1" src="document1" 
               subscribe="transcription" emit="transcription" 
               wait-for="#facsimile #thumbnails"/>
            <pb-facsimile id="facsimile" base-uri="https://apps.existsolutions.com/cantaloupe/iiif/2/" 
               subscribe="transcription"/>
            <pb-image-strip id="thumbnails" base-uri="https://apps.existsolutions.com/cantaloupe/iiif/2/" 
               subscribe="transcription" emit="transcription"
               layout="horizontal" label-position="top"/>
            
         </main>
         ```

**Comment by tuurma** [comment]
Markup example when the list of thumbnails comes from an API endpoint and there's no facsimile viewer as such, thumbnails act as a navigation control

  ```html
         <main>
            <pb-view id="view1" src="document1" 
               subscribe="transcription" emit="transcription" 
               wait-for="#facsimile #thumbnails"/>
            <pb-image-strip id="thumbnails" source="api/thumbnails/{doc}" 
               base-uri="https://apps.existsolutions.com/cantaloupe/iiif/2/"
               load-once="load-once"
               emit="transcription"
               layout="horizontal" label-position="top"/>
         </main>
         ```

**Comment by tuurma** [comment]
Re naming @joern @juri, pb-facsimile already calls it `reference-strip`, following OpenSeaDragon's convention, so I'd perhaps keep it. It's more neutral than facsimile or thumbs.

**Comment by wolfgangmm** [comment]
![image](https://user-images.githubusercontent.com/607433/212893594-5f0108cf-d28f-4073-9b4f-99922d77dd17.png)

`pb-facs-strip` reacts to the same events as `pb-facsimile`. Both components are independent, i.e. they can appear alone on a page and manage their own lists of facsimiles.

## Event handling

* `pb-show-annotation`: as before, is emitted by either `pb-facs-link` (on mouseover) or `pb-facs-strip` to select/display a specific facsimile
* `pb-start-update`: as before, emitted by `pb-view` and `pb-load`. When received by `pb-facsimile` or `pb-facs-strip` this should clear the current list of facsimiles. For example, this way we'll know when the user moves to another page.
* `pb-load-facsimile`: new event emitted by `pb-facs-link`. Causes `pb-facsimile` and `pb-facs-strip` to add the referenced image to their lists.

## Backwards compatibility

Should not an issue as we plug into existing events. Users should not notice the change.

**Comment by juri** [comment]
> Re naming @joern @juri, pb-facsimile already calls it `reference-strip`, following OpenSeaDragon's convention, so I'd perhaps keep it. It's more neutral than facsimile or thumbs.

Wrong Juri. Hi again, @line-o!

**Comment by line-o** [comment]
Moi @juri

**Comment by line-o** [comment]
@wolfgangmm I would assume that the image strip component also listens on `pb-show-annotation` events by others (say: click on a `pb-facs-link`) emitted on the channel its subscribed to in order to highlight the currently shown facsimile.

**Comment by line-o** [comment]
@wolfgangmm Very good schematic BTW. That makes it easier to reason about. We also haven't finished the discussion on how the image strip component is notified of the list of pb-facsimile-links that are present in a `pb-view`.

**Comment by line-o** [comment]
or is that with `pb-load-facsimile`?

**Comment by wolfgangmm** [comment]
@line-o yes, `pb-facs-link` emits `pb-load-facsimile` upon initialization to inform the `pb-facsimile` and/or `pb-facs-strip` about the existance of this image.

And also yes, `pb-facs-strip` could react to `pb-show-annotation` to highlight the current thumb.

**Comment by line-o** [comment]
@duncdrum I believe so
